stats/opencensus/trace.go

```diff
@@ -0,0 +1,165 @@
+/*
+ * Copyright 2022 gRPC authors.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package opencensus
+
+import (
+	"context"
+	"strings"
+	"sync/atomic"
+
+	"go.opencensus.io/trace"
+	"go.opencensus.io/trace/propagation"
+
+	"google.golang.org/grpc/metadata"
+	"google.golang.org/grpc/stats"
+	"google.golang.org/grpc/status"
+)
+
+type spanWithMsgCountKey struct{}
+
+// getSpanWithMsgCount returns the msgEventsCount stored in the context, or nil
+// if there isn't one.
+func getSpanWithMsgCount(ctx context.Context) *spanWithMsgCount {
+	swmc, _ := ctx.Value(spanWithMsgCountKey{}).(*spanWithMsgCount)
+	return swmc
+}
+
+// setSpanWithMsgCount stores a spanWithMsgCount in the context.
+func setSpanWithMsgCount(ctx context.Context, swmc *spanWithMsgCount) context.Context {
+	// needs to be on the heap to persist changes over time so pointer.
+	return context.WithValue(ctx, spanWithMsgCountKey{}, swmc)
+}
+
+// spanWithMsgCount wraps a trace.Span to count msg events.
+type spanWithMsgCount struct {
+	*trace.Span
+	countSentMsg uint32
+	countRecvMsg uint32
+}
+
+// used to attach data client side and deserialize server side
+// traceContextMDKey is the speced key for gRPC Metadata that carries binary
+// serialized traceContext.
+const traceContextMDKey = "grpc-trace-bin"
+
+// traceTagRPC populates context with a new span, and serializes information
+// about this span into gRPC Metadata.
+func (csh *clientStatsHandler) traceTagRPC(ctx context.Context, rti *stats.RPCTagInfo) context.Context {
+	// TODO: get consensus on whether this method name of "s.m" is correct.
+	mn := strings.Replace(removeLeadingSlash(rti.FullMethodName), "/", ".", -1)
+	_, span := trace.StartSpan(ctx, mn, trace.WithSampler(csh.to.TS), trace.WithSpanKind(trace.SpanKindClient))
+
+	swmc := &spanWithMsgCount{
+		Span:         span,
+		countSentMsg: 0, // msg events scoped to scope of context, per attempt client side
+		countRecvMsg: 0,
+	}
+	ctx = setSpanWithMsgCount(ctx, swmc)
+	tcBin := propagation.Binary(span.SpanContext())
+	return metadata.AppendToOutgoingContext(ctx, traceContextMDKey, string(tcBin))
```

Comment:
Ugh, looking at the code, this is what is happening wrt open census tags being sent across the wire in metrics. What you mentioned, using a helper that stores it at the key value pair of the context, rather than the metadata layer, thus that feature will work for a go client/go server for metrics, but not interoperable. Luckily, this addition is orthogonal to the point about the data being used for both traces and stats, but perhaps we could merge these two features in as well? Wdyt about a. the layering I'm talking about and b. the merging of these two features with top level context population?

Reply:
Actually, point b. I realized doesn't apply. We cared about 2 O(n) context linked list look ups because it pulls the data out for each rpc. The tags/span context across the wire doesn't get called per handle call, only on the tag call.

Reply:
We handle these specially in our transport:

https://github.com/grpc/grpc-go/blob/dba26e15a07f43875ccf806a2dd6cbcbc1c12eab/internal/transport/http2_client.go#L566-L571

I think we should be able to use them, and I'd prefer to use accessors for metadata that starts with "grpc" since users aren't actually supposed to be setting it directly themselves.

Reply:
Discussed offline; since the transport handles context k:v -> metadata k:v and vice versa, using these accessors/helpers makes sense on both client and server for both client and server. This will be interoperable with different languages due to the correct metadata k:v being sent across the wire. Will switch this.